lib/puppet/functions/lookup.rb

```diff
@@ -0,0 +1,234 @@
+# has higher priority than data provided in the environment, which has higher priority that data provided in a module,
+#
+# The lookup function can be called in one of these ways:
+#
+#     lookup(name)
+#     lookup(name, value_type)
+#     lookup(name, value_type, merge)
+#     lookup(name, value_type, merge, default_value)
+#     lookup(options_hash)
+#     lookup(name, options_hash)
+#
+# The function may optionally be called with a code block / lambda with the following signature:
+#
+#     lookup(...) |$name| { ... }
+#
+# The block, if present, is mutually exclusive to the `default_value` and will be called with the `name` used in the
+# lookup when no value is found. The value produced by the block then becomes the value produced by the lookup.
+#
+# The meaning of the parameters or content of the options hash is:
+#
+# * `name` - The name or array of names to lookup (first found is returned)
+# * `value_type` - The type to assert. Defaults to 'Data' See 'Type Specification' below.
+# * `default_value` - The default value if there was no value found (must comply with the data type)
+# * `override` - a hash with map from names to values that are used instead of the underlying bindings. If the name
+#   is found here it wins. Defaults to an empty hash.
+# * `default_values_hash` - a hash with map from names to values that are used as a last resort to obtain a value.
+#   Defaults to an empty hash.
+# * `merge` - A string of type Enum[unique,hash,merge] or a hash with the key 'strategy' set to that string. See
+#   'Merge Strategies' below.
+#
+# It is not permitted to pass the `name` as both a parameter and in the options hash.
+#
+# The search will proceed as follows:
+# 1. For each name given in the `name` array (or once, if it's just one name):
+#   - If a matching key is found in the `override` hash, it's value is immediately type checked and returned
+#   - Search and optionally merge Data Binding, environment data providers, and module data providers
+#   - Type check and return the value if a matching key is found
+# 2. For each name given in the `name` array (or once, if it's just one name):
+#   - Type check and return the value if a matching key is found in the `default_values_hash`
+# 3. Type check and return either the given `default_value` or the result of calling the code block if either exist
+# 4. Raise an error indicating that no matching value was found
+#
+# *Merge Strategies*
+#
+# The default behavior of the lookup is to return the first value that is found for the given `name`. The optional
+# `merge` parameter will change this so that a lookup makes an attempt to find values in all three sources (the Data
+# Binder, the environment, and the module scope) and then merge these values according to the given strategy. This
+# does not apply to values found in the 'override' hash. Such values are returned immediately without merging.
+#
```

Comment:
Add "Note that the merge is passed on to allow it to allow the underlying provider to return a merged result"